### PR TITLE
add script for benefit indicator

### PR DESCRIPTION
**Pull request by eperedo**
### :pushpin: References

* **Issue:** Closes https://app.clickup.com/t/865d5j62a

### :memo: Deploy

- [ ] Generate up-to-date input CSV
- [ ] Change default disaggregation of the dataElement (caveat: dataSets will keep the previous category combo)

### :fire: Notes for the reviewer

Docker: **docker.eyeseetea.com/eyeseetea/dhis2-data:2.36.11.1-sp-ip-pro**

Running the script:

```bash
TS_NODE_COMPILER_OPTIONS='{"module":"commonjs"}' npx ts-node src/scripts/benefit-indicator.ts \
--url "http://user:password@localhost:8080" \
--filePath "./B070700-datavalues.csv" \
--csvPath "./report.csv"
```

### CODE REVIEW

File: src/scripts/benefit-indicator.ts [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+
+    await generateReport(dataValuesToSave, orgUnits, metadata, opts.csvPath);
+    console.debug(`Report generated: ${opts.csvPath}`);
+}
+
+async function readCsv(filePath: string): Promise<DataValueCsv[]> {
+    const json = await csvToJson({
+        headers: ["orgUnit", "period", "dataElement", "aocId", "cocId", "value"],
+    }).fromFile(filePath);
+    return json;
```
Comment (tokland) [inline src/scripts/benefit-indicator.ts:81]: Sorry for the confusion, the CSV was intended for the client as a report, not to be used as the script, as we don't to change this specific data values, we want all the data values of that data element. Anyway, let's keep it as is, only as a note, we need to generate the CSV at the moment the script is run, to make sure we are updating everything:

```
{ echo "orgUnit,period"; curl -u "$IP_ADMIN_AUTH" 'https://platform.samaritanspurse.org/api/dataValueSets.json?orgUnit=AGZEUf9meZ6&children=true&startDate=1950&endDate=2050&dataElementGroup=ARzxqc0lEmb' | jq '.dataValues[] | [.orgUnit, .period] | @csv' -r; } > in.csv
```

And change orgUnit filter from code to id.

File: src/scripts/benefit-indicator.ts [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+                orgUnitId,
+                categoryOptionComboId: defaultCocId,
+                attributeOptionComboId: actualCocId,
+                value: String(
+                    _(currentDv).sumBy(dv =>
+                        dv.attributeOptionComboId === actualCocId ? Number(dv.value) : 0
+                    )
+                ),
+                deleted: false,
+            };
```
Comment (tokland) [inline src/scripts/benefit-indicator.ts:307]: [no need to change] The other way was to group DVs also by the AOC.

**Review by tokland — COMMENTED** [review]
I'll add a note on the PR with the steps for deploy.

We are not updating the metadata (dataElement, dataSet->dataSetElements->categoryCombo), should the script do it? The DE is trivial in the UI, the relationships with the dataSet is not. Confirm with @adrianq

**Review by tokland — APPROVED** [review]

**Comment by eperedo** [comment]
> @eperedo As @tokland mentioned, we need to update the datasets. I thought that by simply updating the data element, the dataset will just work. However, the category combo seems to be stored in the data set: image
> 
> Example of a dataset using this data element: https://platform-test.samaritanspurse.org/api/29/dataSets/CKEJAgAF5xl.json?fields=:owner
> 
> Therefore, we have to iterate over all the datasets using this data elements and set the category combo used by this data element to be the 'default' one.

Done

File: src/scripts/benefit-indicator.ts [hunk 3]
```diff
@@ -32,6 +32,12 @@ async function main() {
 
     const { api } = await getApp({ baseUrl: opts.url });
     const metadata = await getMetadata(api);
+
+    const defaultCategoryCombo = metadata.categoryCombos.find(
+        categoryCombo => categoryCombo.name === DEFAULT_CATEGORY_COMBO_NAME
```
Comment (tokland) [inline src/scripts/benefit-indicator.ts:37]: [no need to change] typically we use the code, but the name is always default also.

File: src/scripts/benefit-indicator.ts [hunk 4]
```diff
@@ -227,11 +320,37 @@ async function getMetadata(api: D2Api): Promise<Metadata> {
...
+
+    const d2DataSetResponse = await api
+        .request<{ dataSets: D2DataSet[] }>({
+            method: "get",
+            url: "/dataSets",
+            params: {
+                fields: "id,dataSetElements",
+                filter: `dataSetElements.dataElement.id:eq:${DE_MINISTRY_VISIT_ID}`,
+                paging: false,
+            },
```
Comment (tokland) [inline src/scripts/benefit-indicator.ts:345]: No need to change it, but what was the problem with api.models.dataSets or api.metadata.get({dataSets: ...}) ?

**Review by tokland — APPROVED** [review]
All good, categoryCombo is updated for the dataSets.

File: src/scripts/benefit-indicator.ts (hunk 4 above)
Comment (eperedo) [reply src/scripts/benefit-indicator.ts:345]: I was trying to apply the filter using `metadata.get`, but typescript was complaining.

```
filter: {
  dataSetElements: {
    dataElement: {
      eq: DE_MINISTRY_VISIT_ID
    }
 }
}
```

Looks like we cannot add filter for nested fields?

**Pull request merged**